### PR TITLE
SASS-1961 : Added Tax Year List Retrieval connectors and Tax Year Act…

### PR DESCRIPTION
**Pull request by robert-webb-hmrc**
…ion Function

### Description
Please include a brief summary of the change you've made and the issue that it fixes.

Add a link to the relevant story in Jira
[SASS-1961](https://jira.tools.tax.service.gov.uk/browse/SASS-1961)

##### Before Reviewing
- [x]  Have you pulled the branch down?
- [x]  Have you assigned yourself to the PR?
- [x]  Have you moved the task to “in review” on JIRA?
- [x]  Have you checked to ensure all dependencies are up to date?
- [x]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [x]  Have you run the tests?
- [x]  Have you run the journey tests?
- [x]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [x]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [x]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [x]  Have you run the journey tests? (where applicable)
- [x]  Have you addressed warnings where appropriate?
- [x]  Have you rebased against the current version of main?
- [x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [x]  Have you checked the PR Builder passes?

### CODE REVIEW

File: app/controllers/DeclarationPageController.scala [hunk 1]
```diff
@@ -42,7 +43,8 @@ class DeclarationPageController @Inject()(declareCrystallisationService: Declare
                                           implicit val ec: ExecutionContext,
                                           declarationPageView: DeclarationPageView,
                                           authorisedAction: AuthorisedAction,
-                                          errorHandler: ErrorHandler,
+                                          implicit val validTaxYearListConnector: ValidTaxYearListConnector,
```
Comment (BenCosfordHMRC) [inline app/controllers/DeclarationPageController.scala:46]: we shouldn't be supplying connectors to controllers, only services. Can swap out to pass the service in instead.

File: app/controllers/OverviewPageController.scala [hunk 2]
```diff
@@ -37,7 +38,8 @@ class OverviewPageController @Inject()(inYearAction: InYearAction,
                                        liabilityCalculationService: LiabilityCalculationService,
                                        overviewPageView: OverviewPageView,
                                        authorisedAction: AuthorisedAction,
-                                       errorHandler: ErrorHandler,
+                                       implicit val validTaxYearListConnector: ValidTaxYearListConnector,
```
Comment (BenCosfordHMRC) [inline app/controllers/OverviewPageController.scala:41]: as above

File: app/controllers/predicates/TaxYearAction.scala [hunk 3]
```diff
@@ -16,21 +16,25 @@
...
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
 
 import javax.inject.Inject
 import scala.concurrent.{ExecutionContext, Future}
 
 class TaxYearAction @Inject()(taxYear: Int, missingTaxYearReset: Boolean)(
   implicit val appConfig: AppConfig,
-  val messagesApi: MessagesApi
+  val messagesApi: MessagesApi,
+  validTaxYearListConnector: ValidTaxYearListConnector,
```
Comment (BenCosfordHMRC) [inline app/controllers/predicates/TaxYearAction.scala:36]: as above can swap this to the service instead and do the same thing.

**Review by BenCosfordHMRC — CHANGES_REQUESTED** [review]

File: app/controllers/DeclarationPageController.scala (hunk 1 above)
Comment (robert-webb-hmrc) [reply app/controllers/DeclarationPageController.scala]: Updated

**Review by BenCosfordHMRC — APPROVED** [review]

File: app/controllers/predicates/TaxYearAction.scala [hunk 4]
```diff
@@ -39,29 +43,54 @@ class TaxYearAction @Inject()(taxYear: Int, missingTaxYearReset: Boolean)(
...
-    Future.successful(
-      if (!appConfig.taxYearErrorFeature || taxYear == appConfig.defaultTaxYear) {
-        val sameTaxYear = request.session.get(TAX_YEAR).exists(_.toInt == taxYear)
+    implicit lazy val headerCarrier: HeaderCarrier = HeaderCarrierConverter.fromRequestAndSession(request,request.session)
 
+    def taxYearListCheck(sessionExists: Boolean, validTaxYears: Seq[Int]) = {
+      if (!appConfig.taxYearErrorFeature || validTaxYears.contains(taxYear)) {
+        val sameTaxYear = request.session.get(TAX_YEAR).exists(_.toInt == taxYear)
         if (sameTaxYear || !missingTaxYearReset) {
           Right(request)
```
Comment (BenCosfordHMRC) [inline app/controllers/predicates/TaxYearAction.scala:52]: think we need to add valid tax years on the same tax year line too so llike

`if(sessionExists){
            Right(request)
          } else {
            val newSessionData = Seq(TAX_YEAR -> taxYear.toString, VALID_TAX_YEARS -> validTaxYears.mkString(","))
            Left(Redirect(controllers.routes.StartPageController.show(taxYear).url).addingToSession(newSessionData: _*))
          }`
          
been testing it with employment FE and needs the valid tax years to be in session and this seems to fix it

**Review by BenCosfordHMRC — APPROVED** [review]

**Pull request merged**